File: templates/problems/05_binary_search/06_time_based_kv/time_based_kv_test.go
package time_based_kv

import "testing"

func TestTimeMap(t *testing.T) {
	tests := []struct {
		name       string
		operations []string
		keys       []string
		values     []string
		timestamps []int
		expected   []string
	}{
		{
			name:       "basic set and get",
			operations: []string{"Set", "Get", "Get", "Set", "Get", "Get"},
			keys:       []string{"foo", "foo", "foo", "foo", "foo", "foo"},
			values:     []string{"bar", "", "", "bar2", "", ""},
			timestamps: []int{1, 1, 3, 4, 4, 5},
			expected:   []string{"", "bar", "bar", "", "bar2", "bar2"},
		},
		{
			name:       "get before any set",
			operations: []string{"Get"},
			keys:       []string{"foo"},
			values:     []string{""},
			timestamps: []int{1},
			expected:   []string{""},
		},
		{
			name:       "get with timestamp before first set",
			operations: []string{"Set", "Get"},
			keys:       []string{"key", "key"},
			values:     []string{"val", ""},
			timestamps: []int{5, 3},
			expected:   []string{"", ""},
		},
		{
			name:       "multiple keys",
			operations: []string{"Set", "Set", "Get", "Get"},
			keys:       []string{"a", "b", "a", "b"},
			values:     []string{"v1", "v2", "", ""},
			timestamps: []int{1, 1, 1, 1},
			expected:   []string{"", "", "v1", "v2"},
		},
		{
			name:       "get exact timestamp",
			operations: []string{"Set", "Set", "Set", "Get", "Get", "Get"},
			keys:       []string{"k", "k", "k", "k", "k", "k"},
			values:     []string{"a", "b", "c", "", "", ""},
			timestamps: []int{1, 2, 3, 1, 2, 3},
			expected:   []string{"", "", "", "a", "b", "c"},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			tm := NewTimeMap()
			for i, op := range tt.operations {
				switch op {
				case "Set":
					tm.Set(tt.keys[i], tt.values[i], tt.timestamps[i])
				case "Get":
					if got := tm.Get(tt.keys[i], tt.timestamps[i]); got != tt.expected[i] {
						t.Errorf("step %d: Get(%q, %d) = %q, want %q", i, tt.keys[i], tt.timestamps[i], got, tt.expected[i])
					}
				}
			}
		})
	}
}

File: templates/problems/05_binary_search/06_time_based_kv/typescript/time_based_kv.test.ts
import { describe, it, expect } from "vitest";
import { TimeMap } from "./time_based_kv";

describe("TimeMap", () => {
  it("basic set and get", () => {
    const tm = new TimeMap();
    tm.set("foo", "bar", 1);
    expect(tm.get("foo", 1)).toBe("bar");
    expect(tm.get("foo", 3)).toBe("bar");
    tm.set("foo", "bar2", 4);
    expect(tm.get("foo", 4)).toBe("bar2");
    expect(tm.get("foo", 5)).toBe("bar2");
  });

  it("get before any set", () => {
    const tm = new TimeMap();
    expect(tm.get("foo", 1)).toBe("");
  });

  it("get with timestamp before first set", () => {
    const tm = new TimeMap();
    tm.set("key", "val", 5);
    expect(tm.get("key", 3)).toBe("");
  });

  it("multiple keys", () => {
    const tm = new TimeMap();
    tm.set("a", "v1", 1);
    tm.set("b", "v2", 1);
    expect(tm.get("a", 1)).toBe("v1");
    expect(tm.get("b", 1)).toBe("v2");
  });

  it("get exact timestamp", () => {
    const tm = new TimeMap();
    tm.set("k", "a", 1);
    tm.set("k", "b", 2);
    tm.set("k", "c", 3);
    expect(tm.get("k", 1)).toBe("a");
    expect(tm.get("k", 2)).toBe("b");
    expect(tm.get("k", 3)).toBe("c");
  });
});
